### ISSUE TITLE
training with mosaic() and negative sample questions 

### ISSUE THREAD

**Issue by jamjamjon** [ultralytics/yolov5#6304]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi, sorry to interrupt you. I met some problem about mosaic skills when training.

here is the thing.  I recently have a project which is to detect people who ride a motorcycle,  not to detect motor or people individually, but detect them all, as a whole. I've used Darknet YOLOv4 and found the result is not good, so I decided to use YOLOv5 to train it. My datasets has 10,000+ images, which contains 1/5 images has only people or only motorcycle or bycicle as negative samples. Those negative images’ label file are empty .txt.  

Training results is great but has some strange phenomenon: some image only has people or motorcycle, and the people or motorcycle is detected individually. This is surely not what I want. I have checked the training datasets and the labels is right, but using the trained model to infer the negative sample images in training datasets, some of the images would be detected with only people or motorcycle.

Hyp in training is using med.yaml. Mosaic =1.0 .

 Then I check the load_mosaic() function and found out some segments in img4 generated by 4 images after mosaic() will be covered or hide. Therefore,   image(man ride on a motorcycle ) after mosaic would probably lost the front of the object(people and part of motorcycle), only the back of the motorcycle will be one the img4 used for training. I think this the main reason caused this problem. 

I want to known how to fix this problem except using -rect or not using mosaic directly. Besides, will it work of the way I do with negative sample images which just use empty .txt label file?  And should I adjust the hyper-parameters in hyp-med.yaml to deal with negative  sAmples?

My English is not good, if I did not express clear, please let me know. 

Really looking forward for your suggestion. Thanks!

### Additional

original image :
<img width="593" alt="original-image" src="https://user-images.githubusercontent.com/51357717/149606098-4a260c3c-2863-4fe7-9b2d-86b4e13c5da2.png">

after mosaic()  might be:
<img width="354" alt="after-mosaic-might-be" src="https://user-images.githubusercontent.com/51357717/149606117-55871bee-6831-48c9-a90d-88ecc7f87fbe.png">

**Comment by zhiqwang** [comment]
Hi @Jamjamjamjon ,

Why not just use a classification model to do this task, and would a detection box be particularly helpful for your scenario?

**Comment by glenn-jocher** [comment]
@Jamjamjamjon I think you might be confusing augmentations. Mosaic will join 4 images together and then crop the outside to maintain imgsz. That's it. What you are showing is one image with `scale` and `translate` augmentation. Scale, translate and mosaic are 3 separate augmentations that are independent of each other and can be modified in your hyp file:

https://github.com/ultralytics/yolov5/blob/436ffc417ac2312de18287ddc4f87bdc2f7f5734/data/hyps/hyp.scratch.yaml#L25-L34

About reducing FPs, you just need more background images and to make sure you train enough that overtraining occurs. A full list of recommendations for improving results is below.

👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by glenn-jocher** [comment]
@Jamjamjamjon also you can verify your labels by examining your train*.jpg batches generated on training start.

**Comment by jamjamjon** [comment]
> @Jamjamjamjon also you can verify your labels by examining your train*.jpg batches generated on training start.

I found the mistake! I double check the training datasets and found out that there is a Negative sample image has been labeled! I haven’t finish examining all the dataset, I guess there would be several mistakes like like still exist!
Thanks for your reply! Have a good day!

**Comment by njustczr** [comment]
i think you should set translate=0.0

**Comment by glenn-jocher** [comment]
@njustczr thanks for the suggestion! We recommend users first train with all default settings before considering any changes, but adjusting the `translate` parameter might be something you could experiment with. Feel free to give it a try and let us know how it goes! If you have any other questions, feel free to ask. Good luck!
